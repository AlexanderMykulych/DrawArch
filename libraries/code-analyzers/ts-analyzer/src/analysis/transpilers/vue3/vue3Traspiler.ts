import { parse, compileScript, compileTemplate } from 'vue/compiler-sfc'
import type { TranspilerResult } from '../types'

export default {
  name: 'vue3-traspiler',
  async traspile(code: string): Promise<TranspilerResult> {
    const { descriptor } = parse(code, { sourceMap: true })

    const compiled = descriptor.script || descriptor.scriptSetup ? compileScript(descriptor, {
      id: `1`,
      sourceMap: true,
    }) : null

    const template = descriptor.template?.content ? compileTemplate({
      source: descriptor.template.content,
      filename: 'test.vue',
      id: 'test.vue',
      compilerOptions: {
        sourceMap: true,
      }
    }) : null
    
    const result = `${compiled?.content ?? '// no <script> tag'}\n${template?.code}`

    return Promise.resolve({
      code: result,
      sourceMaps: compiled?.map,
    })
  },
}
