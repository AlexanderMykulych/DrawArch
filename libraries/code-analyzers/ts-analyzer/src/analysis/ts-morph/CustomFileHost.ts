import { FileSystemHost, RuntimeDirEntry } from "ts-morph";
import { ViteDevServer } from "vite";

export class CustomFileHost implements FileSystemHost {
  viteDevServer: ViteDevServer;
  baseFileHost: FileSystemHost;
  constructor(viteDevServer: ViteDevServer, baseFileHost: FileSystemHost) {
    this.viteDevServer = viteDevServer
    this.baseFileHost = baseFileHost
  }
  isCaseSensitive(): boolean {
    return this.baseFileHost.isCaseSensitive()
  }
  delete(path: string): Promise<void> {
    return this.baseFileHost.delete(path)
  }
  deleteSync(path: string): void {
    this.baseFileHost.deleteSync(path)
  }
  readDirSync(dirPath: string): RuntimeDirEntry[] {
    return this.baseFileHost.readDirSync(dirPath)
  }
  async readFile(filePath: string, encoding?: string | undefined): Promise<string> {
    const code = await this.baseFileHost.readFile(filePath)
    const result = await this.viteDevServer.pluginContainer.transform(code, filePath)

    return result?.code ?? code
  }
  readFileSync(filePath: string, encoding?: string | undefined): string {
    throw new Error("Method not implemented.");
  }
  writeFile(filePath: string, fileText: string): Promise<void> {
    return this.baseFileHost.writeFile(filePath, fileText)
  }
  writeFileSync(filePath: string, fileText: string): void {
    this.baseFileHost.writeFileSync(filePath, fileText)
  }
  mkdir(dirPath: string): Promise<void> {
    return this.baseFileHost.mkdir(dirPath)
  }
  mkdirSync(dirPath: string): void {
    this.baseFileHost.mkdirSync(dirPath)
  }
  move(srcPath: string, destPath: string): Promise<void> {
    return this.baseFileHost.move(srcPath, destPath)
  }
  moveSync(srcPath: string, destPath: string): void {
    this.baseFileHost.moveSync(srcPath, destPath)
  }
  copy(srcPath: string, destPath: string): Promise<void> {
    return this.baseFileHost.copy(srcPath, destPath)
  }
  copySync(srcPath: string, destPath: string): void {
    this.baseFileHost.copySync(srcPath, destPath)
  }
  fileExists(filePath: string): Promise<boolean> {
    return this.baseFileHost.fileExists(filePath)
  }
  fileExistsSync(filePath: string): boolean {
    throw new Error("Method not implemented.");
  }
  directoryExists(dirPath: string): Promise<boolean> {
    return this.baseFileHost.directoryExists(dirPath)
  }
  directoryExistsSync(dirPath: string): boolean {
    return this.baseFileHost.directoryExistsSync(dirPath)
  }
  realpathSync(path: string): string {
    return this.baseFileHost.realpathSync(path)
  }
  getCurrentDirectory(): string {
    return this.baseFileHost.getCurrentDirectory()
  }
  glob(patterns: readonly string[]): Promise<string[]> {
    return this.baseFileHost.glob(patterns)
  }
  globSync(patterns: readonly string[]): string[] {
    return this.baseFileHost.globSync(patterns)
  }

}