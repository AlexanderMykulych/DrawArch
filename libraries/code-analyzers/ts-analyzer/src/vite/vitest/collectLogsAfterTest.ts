import { ActionEntity, ActionEntityType, CodeInfo, CodeInfoType, Relationship, RelationType } from "../../analysis/types"
import { useNeo4j } from "../../neo4j/useNeo4j"
import type { LogItem, StartLogItem } from "../shim"


export async function collectLogsAfterTest(insertToDb: true): Promise<void>
export async function collectLogsAfterTest(): Promise<CodeInfo[]>
export async function collectLogsAfterTest(insertToDb?: true): Promise<CodeInfo[] | void> {
  if (!globalThis.__getLogs) {
    return
  }

  const logs = globalThis.__getLogs()

  const codeInfos = logsToCodeInfo(logs)

  if (insertToDb) {
    const db = useNeo4j()
    await db.insertCodeInfos(codeInfos);
  } else {
    return codeInfos
  }
}


export function logsToCodeInfo(logs: LogItem[]): CodeInfo[] {
  return logs
    .flatMap<CodeInfo>((item, index, items) => {
      const result = []
      if (item.type === 'end') {
        const callEntity: ActionEntity = {
          type: CodeInfoType.Entity,
          entityType: ActionEntityType.Call,
          id: getCallActionId(item, index),
          executionTime: item.execTime,
        }
        const callRelation: Relationship = {
          type: CodeInfoType.Relationship,
          relationType: RelationType.CallEntity,
          from: callEntity.id,
          to: item.id,
        }

        result.push(callEntity, callRelation)

        const [parentStartItem, parentIndex] = findParentStartLogFromIndex(items, index, item.id)

        if (parentStartItem) {
          const parentRelation: Relationship = {
            type: CodeInfoType.Relationship,
            relationType: RelationType.Parent,
            from: getCallActionId(parentStartItem, parentIndex),
            to: callEntity.id,
          }
          result.push(parentRelation)
        }
      }
      return result
    })
}

export function findStartLogFromIndex(items: LogItem[], index: number, id: string): StartLogItem {
  for (let i = index - 1; i >= 0; i--) {
    const startItem = items[i]
    if (startItem.type === 'start' && startItem.id === id) {
      return startItem
    }
  }
  throw 'Can not find start element for: ' + id
}

export function findParentStartLogFromIndex(items: LogItem[], index: number, id: string): [StartLogItem, number] | [null, null] {
  for (let i = index - 1; i >= 0; i--) {
    const startItem = items[i]
    if (startItem.type === 'start' && startItem.id !== id) {
      return [startItem, i]
    }
  }

  return [null, null]
}

export function getCallActionId(item: LogItem, index: number): string {
  // return `call:${item.id}`
  return `call[${index}]:${item.id}`
}