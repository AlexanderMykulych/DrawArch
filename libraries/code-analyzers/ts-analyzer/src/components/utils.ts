import { Node } from "ts-morph";

export function getNodeName(node: Node | undefined): string | undefined {
  if (Node.isVariableDeclaration(node)) {
    return node.getName()
  }
  if (Node.isFunctionDeclaration(node)) {
    return node.getName()
  }
  if (Node.isParameterDeclaration(node)) {
    return node.getName()
  }
  if (Node.isExportSpecifier(node)) {
    return node.getName()
  }
}