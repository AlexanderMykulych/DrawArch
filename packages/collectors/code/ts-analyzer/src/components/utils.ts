import { Node } from "ts-morph";

export function getNodeName(node: Node | undefined): string | undefined {
  if (Node.isVariableDeclaration(node) 
    || Node.isFunctionDeclaration(node)
    || Node.isParameterDeclaration(node)
    || Node.isMethodDeclaration(node)
    || Node.isExportSpecifier(node)
  ) {
    return node.getName()
  }
}