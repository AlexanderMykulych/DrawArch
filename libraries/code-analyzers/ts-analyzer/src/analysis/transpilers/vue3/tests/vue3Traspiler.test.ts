import { Project } from "ts-morph";
import { expect, test } from "vitest";
import vue3Traspiler from "../vue3Traspiler";

const vueCode = `
<script lang="ts" setup>
import { ref } from 'vue'
import { utilFn } from './utils'

const count = ref(utilFn(1, 2))
</script>

<template>
<div>
  {{ count }}
</div>
</template>
`
test('should transpile vue file to ts with default export and render fn export', async () => {
  const tsCode = await vue3Traspiler.traspile(vueCode)

  const project = new Project({
    useInMemoryFileSystem: true,
    skipFileDependencyResolution: true,
  })
  const sourceFile = project.createSourceFile('component.vue.ts', tsCode)

  const exportDeclarations = sourceFile.getExportedDeclarations()

  expect(exportDeclarations).toHaveLength(2)
  expect(exportDeclarations.get('render')).not.empty
  expect(exportDeclarations.get('default')).not.empty
})

test('should transpile vue code to correct TS code', async () => {
  const tsCode = await vue3Traspiler.traspile(vueCode)

  const project = new Project({
    useInMemoryFileSystem: true,
    skipFileDependencyResolution: true,
  })
  const fs = project.getFileSystem()

  await fs.writeFile('/node_modules/vue/index.js', 'export const ref = () => {}')
  await fs.writeFile('/node_modules/vue/index.d.ts', `
  export const ref: any
  export const defineComponent: any
  export const toDisplayString: any
  export const openBlock: any
  export const createElementBlock: any
  `)
  project.createSourceFile('utils.ts', `export const utilFn = (a, b) => {}`)
  const sourceFile = project.createSourceFile('component.vue.ts', tsCode)

  const diagnostincs = sourceFile.getPreEmitDiagnostics()


  console.log(diagnostincs.map(x => x.getMessageText()))
  expect(diagnostincs).is.empty
})
