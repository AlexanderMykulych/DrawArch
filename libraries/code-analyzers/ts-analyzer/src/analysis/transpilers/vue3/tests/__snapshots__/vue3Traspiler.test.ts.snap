// Vitest Snapshot v1

exports[`should generate source-map 1`] = `
{
  "column": 10,
  "line": 3,
  "name": null,
  "source": "anonymous.vue",
}
`;

exports[`should transpile vue file to ts with default export and render fn export 1`] = `
{
  "code": "import { defineComponent as _defineComponent } from 'vue'
import { ref } from 'vue'
import { utilFn } from './utils'


export default /*#__PURE__*/_defineComponent({
  setup(__props, { expose }) {
  expose();

const count = ref(utilFn(1, 2))

const __returned__ = { count }
Object.defineProperty(__returned__, '__isScriptSetup', { enumerable: false, value: true })
return __returned__
}

})
import { toDisplayString as _toDisplayString, openBlock as _openBlock, createElementBlock as _createElementBlock } from \\"vue\\"

export function render(_ctx, _cache) {
  return (_openBlock(), _createElementBlock(\\"div\\", null, _toDisplayString(_ctx.count), 1 /* TEXT */))
}",
  "sourceMaps": SourceMap {
    "file": null,
    "mappings": ";AAEA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACzB,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AAChC;;;;;AAHwB;AAIxB,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;;;;;;;",
    "names": [],
    "sources": [
      "anonymous.vue",
    ],
    "sourcesContent": [
      "
<script lang=\\"ts\\" setup>
import { ref } from 'vue'
import { utilFn } from './utils'

const count = ref(utilFn(1, 2))
</script>

<template>
<div>
  {{ count }}
</div>
</template>
",
    ],
    "version": 3,
  },
}
`;
